File: rb/mm-native.rb
require 'matrix'

class File
  def read_toks; readline.strip.split; end
  def read_ints; read_toks.map(&:to_i); end
  def read_floats; read_toks.map(&:to_f); end
  def read_matrix rows; Matrix[*rows.times.collect { read_floats }]; end
end

if ARGV.empty?
  puts "Usage: #{$0} <input>"
  exit false
end

# Read matricies from file.
file = ARGV.first
a, b = File.open(file) do |fp|
  m, k, n = fp.read_ints
  [fp.read_matrix(m), fp.read_matrix(k)]
end

# Multiply.
c = a * b

File: py/mm-native.py
from __future__ import print_function
from sys import argv, stderr, exit

def read_toks(fp):
    return fp.readline().split()
def read_ints(fp):
    return map(int, read_toks(fp))
def read_floats(fp):
    return map(float, read_toks(fp))
def read_matrix(fp, rows):
    return [read_floats(fp) for i in range(rows)]

if len(argv) != 2:
    print("Usage: {} <input>".format(argv[0]),
            file=stderr)
    exit(1)

filename = argv[1]
with open(filename) as fp:
    m, k, n = read_ints(fp)
    a = read_matrix(fp, m)
    b = read_matrix(fp, k)

c = [[0 for i in range(n)] for i in range(m)]
for i in range(m):
    for j in range(k):
        for h in range(n):
            c[i][j] += a[i][h] * b[h][j]
